power: add system power management direct trigger mode

wentongwu opened the pull request:
Add system power management direct trigger mode. In this mode
application thread can directly put system in sleep or deep
sleep mode instead of waiting for idle thread to do it, so that
it can reduce latency to enter low power mode.

Fixes: #20125.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu commented on subsys/power/power.c:
```diff
@@ -81,6 +87,10 @@ void sys_pm_force_power_state(enum power_states state)
 
 	forced_pm_state = state;
+
+#ifdef CONFIG_SYSTEM_PM_WORKQUEUE
```
@vanti I believe there is the case triggering system suspend in interrupt context where maybe _sys_suspend isn't suitable to execute. And also the latency of idle thread is from waiting other threads suspend, and as usual the workqueue will be in high priority, so maybe the latency of workqueue will be one context switch.

vanti replied:
The context of the caller of sys_pm_force_power_state() will be running when the function is invoked, so there won't be any waiting for the other threads to suspend, regardless of the caller's priority. And given you are disabling interrupts prior to calling _sys_suspend(), there is no latency from servicing interrupts either. Thus a direct call to _sys_suspend() would be the fastest way to get to suspend, with zero context switch.

For the users who just want to suspend the system from a thread context, bringing in workqueues seems like a significant penalty. I'd suggest to either make the workqueue an optional alternative to a direct call to _sys_suspend(), or (better yet) remove it and simply let the application create its own thread to call sys_pm_force_power_state(), that the interrupt context would make ready if suspend is desired. It would just be standard "bottom-half" interrupt handling, and the application can prioritize the thread as highly as necessary.

wentongwu committed:
Style: remove extern on the function declarations in the header

Remove extern on the function declarations in the header file to
keep consistent with others.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

vanti reviewed (changes requested):
Thanks for the rework. I think if you document that this API should be called in thread context when direct mode is set, it would make it clear to the users how to use it.

wentongwu commented:
> Thanks for the rework. I think if you document that this API should be called in thread context when direct mode is set, it would make it clear to the users how to use it.

@vanti yes, we should update document, please review the new patch.

vanti commented on subsys/power/power.c:
```diff
@@ -152,6 +158,12 @@ enum power_states _sys_suspend(s32_t ticks)
 	if (!post_ops_done) {
 		post_ops_done = 1;
+
+#ifdef CONFIG_SYS_PM_DIRECT_FORCE_MODE
```
What is the motivation behind having a different behavior here for direct force mode? If there is a need to clear the forced state, maybe it would apply for the standard mode as well. I think it is simpler to keep the behavior consistent across the two modes, unless there is a good reason behind it.

wentongwu replied:
@vanti the original documentation says this function will impact all subsequent
 suspend operations until forcing the SYS_POWER_STATE_AUTO state restores normal operation. I'm not clear what's the usage for this operation but I keep it as original design.
```
/**
 * @brief Force usage of given power state.
 *
 * This function overrides decision made by PM policy
 * forcing usage of given power state in all subsequent
 * suspend operations. Forcing the SYS_POWER_STATE_AUTO
 * state restores normal operation.
 *
 * When enabled SYS_PM_DIRECT_FORCE_MODE, this function
 * can only run in thread context. And before the end of
 * the function forced_pm_state is cleared, so it doesn't
 * impact the subsequent suspend operations in this mode.
 *
 * @param state Power state which should be used in all
 *              subsequent suspend operations or the ongoing
 *              suspend operation in direct force mode or
 *              SYS_POWER_STATE_AUTO.
 */
void sys_pm_force_power_state(enum power_states state);
```

vanti replied:
If we choose to keep the behavior backward-compatible, I think we should stick to it in direct force mode as well (i.e. remove this code block). Right now it looks error-prone - if I started with code that calls sys_pm_force_power_state() using the standard mode, and decide to throw CONFIG_SYS_PM_DIRECT_FORCE_MODE to reduce the latency (which I think is the purpose of the option), all of a sudden the forced state is automatically cleared. This is rather unexpected.

wentongwu replied:
@vanti if there is no special use case to force power state in all subsequent suspend operations, I will  change this API to impact only one-shot.

vanti replied:
If that is the behavior you want, it'd be good to submit it as a separate pull request, and introduce the behavior for all cases. Another possibility is to have two different API: sys_pm_force_power_state() and sys_pm_force_power_state_one_shot() to differentiate the two if you wish to support both behavior.

wentongwu replied:
@vanti ok, let's make the decision here and now that sys_pm_force_power_state will only work for one-shot.

vanti replied:
I am ok with the change in behavior, though would have made the change into at least a separate commit, given it is a different change from direct force mode. Let's see if others have comments regarding it.

wentongwu replied:
@vanti I add one more commit to separate them. Thanks

wentongwu committed:
Power: change sys_pm_force_power_state only works for one shot

Change sys_pm_force_power_state only works for the current ongoing
suspend operation, before the end of syspend state forced_pm_state
will be cleared.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
power: add system power management direct force mode.

Add system power management direct force trigger mode. In this
mode application thread can directly put system in sleep or deep
sleep mode instead of waiting for idle thread to do it, so that
it can reduce latency to enter low power mode.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

MZachmann reviewed (commented):
Am I confused? As far as I can tell this doesn't compile because on line 86 of power.c _sys_suspend takes an s32_t not a k_timeout argument.

erwango commented:
@MZachmann , can you raise an issue and detail the configuration in which you're seeing this issue?

MZachmann commented:
done. see https://github.com/zephyrproject-rtos/zephyr/issues/25354
